### PR TITLE
Workaround for P/Invoke AccessViolationException

### LINKED ISSUE #415: 'System.AccessViolationException' in Facebook.Yoga.dll
The app is running fine in simulator by react-native run-windows, but it will not run on a device through Visual Studio 2015 DebugRelease ARM. I have run the packaging command as per the docs.

It seems an issue with Facebook Yoga which is tied in with https://github.com/ReactWindows/react-native-windows, but I think this is the best place to report?

Running facebook.yoga version 1.1.0.18-pre on the ReactNative package (not the app package).

Exception thrown: 'System.AccessViolationException' in Facebook.Yoga.dll
The thread 0x92c has exited with code 0 (0x0).
The thread 0x11d4 has exited with code 0 (0x0).
The thread 0x10cc has exited with code 0 (0x0).
The program '[4644] Name.exe' has exited with code 0 (0x0).

**Edit**
Please note this is trying to run set to ARM for a Windows 10 mobile, not in X64 mode

Any thoughts?
Vince

(I tried to also install via the new npm but I am getting lots of errors:)

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "--" "run" "copy-sources"
npm ERR! node v7.5.0
npm ERR! npm  v4.1.2
npm ERR! code ELIFECYCLE
npm ERR! yoga-layout@1.0.1 copy-sources: `! npm -- run is-monolithic || (rsync -r --checksum --delete ../yoga/ sources/yoga/)`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the yoga-layout@1.0.1 copy-sources script '! npm -- run is-monolithic || (rsync -r --checksum --delete ../yoga/ sources/yoga/)'.
npm ERR! Make sure you have the latest version of node.js and npm installed.

### PR DESCRIPTION
**Pull request by rozele**
The issue is on ARM builds for Windows UWP. For the C# P/Invoke API wrapper, any native method that returns the YogaValue struct throws the AccessViolationException. The issue is not with structs in general, as returning the YogaSize / YGSize struct works fine. The issue seems to be limited to structs that have an enum member.

I tried a number of things to resolve the issue without changing the underlying native API for Windows. I read the ARM documentation and saw reference to variable enum sizes based on the number of enum members, so I tried to use a number of different UnmanagedType values in a [MarsalAs()] attribute on the enum member of the C# struct declaration.

What ultimately worked was to return a pointer to the location of the struct, and use the System.Runtime.InteropServices.PtrToStructure API to read the struct data from that pointer. I added a few new macros that will return the struct address on Windows only, other builds are not affected.

Note, I have not tested the impact of this change on Xamarin.iOS, Xamarin.Mac, or Xamarin.Android, and I suspect some changes will need to be made to accomodate these target frameworks (cc @rmarinho).

The downside to this PR is that System.Runtime.InteropServices.PtrToStructure returns System.Object, which means the YogaValue struct will incur boxing, ultimately having an impact on the heap and Gen0 GC collections. However, I think this is still a reasonable workaround until we can figure out why the default struct marshalling behavior for P/Invoke is throwing the AVE for structs with enums.

Fixes #415

### CODE REVIEW

**Comment by splhack** [comment]
Can we have "#if directive" for ARM?

**Comment by rozele** [comment]
@splhack potentially, do you know any preprocessor directives that are set for ARM builds for the unmanaged code?

**Comment by splhack** [comment]
We don't need to worry about other platforms if we have "#if directive" for ARM builds for Windows UWP.

**Comment by rozele** [comment]
@splhack isolated the changes to UWP ARM, still need to update the .nuspec

**Comment by splhack** [comment]
@rozele sweet! I love it

File: csharp/Facebook.Yoga/YogaNode.Spacing.cs [hunk 1]
```diff
@@ -17,7 +15,7 @@ public YogaValue Left
         {
             get
             {
-                return Native.YGNodeStyleGetPosition(_ygNode, YogaEdge.Left);
+                return StructHelpers.MarshalStruct<YogaValue>(Native.YGNodeStyleGetPosition(_ygNode, YogaEdge.Left));
```
Comment (rmarinho) [inline csharp/Facebook.Yoga/YogaNode.Spacing.cs]: Humm i not very fan of this.. using generics will have a performance hit on Android.

File: csharp/Windows/Facebook.Yoga.Universal/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Reflection;
```
Comment (rmarinho) [inline csharp/Windows/Facebook.Yoga.Universal/Properties/AssemblyInfo.cs]: We can drop this file

File: csharp/Windows/Facebook.Yoga.Universal/Properties/Facebook.Yoga.Universal.rd.xml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (rmarinho) [inline csharp/Windows/Facebook.Yoga.Universal/Properties/Facebook.Yoga.Universal.rd.xml]: I think we can drop this file

File: csharp/Facebook.Yoga/YogaNode.Spacing.cs (hunk 1 above)
Comment (rozele) [inline csharp/Facebook.Yoga/YogaNode.Spacing.cs]: Fixed.

**Comment by rozele** [comment]
@splhack can you re-import, had to change the visibility of the marshaling helper

**Pull request closed without merging**